File: backend/internal/repository/postgres/audit_log_repository.go
package postgres

import (
	"en-inventory/internal/domain"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type auditLogRepository struct {
	db *gorm.DB
}

func NewAuditLogRepository(db *gorm.DB) domain.AuditLogRepository {
	return &auditLogRepository{
		db: db,
	}
}

func (r *auditLogRepository) Create(log *domain.AuditLog) error {
	if log.ID == uuid.Nil {
		log.ID = uuid.New()
	}
	log.CreatedAt = time.Now()

	return r.db.Create(log).Error
}

func (r *auditLogRepository) List(filter map[string]interface{}) ([]domain.AuditLog, error) {
	var logs []domain.AuditLog

	query := r.db.Model(&domain.AuditLog{})

	// Apply filters
	if entityType, ok := filter["entity_type"]; ok {
		query = query.Where("entity_type = ?", entityType)
	}
	if entityID, ok := filter["entity_id"]; ok {
		query = query.Where("entity_id = ?", entityID)
	}
	if action, ok := filter["action"]; ok {
		query = query.Where("action = ?", action)
	}
	if userID, ok := filter["user_id"]; ok {
		query = query.Where("user_id = ?", userID)
	}
	if fromDate, ok := filter["from_date"]; ok {
		query = query.Where("created_at >= ?", fromDate)
	}
	if toDate, ok := filter["to_date"]; ok {
		query = query.Where("created_at <= ?", toDate)
	}

	err := query.Order("created_at DESC").Find(&logs).Error
	return logs, err
}

func (r *auditLogRepository) ListPaginated(filter map[string]interface{}, page, pageSize int) ([]domain.AuditLog, int, error) {
	var logs []domain.AuditLog
	var total int64

	query := r.db.Model(&domain.AuditLog{})

	// Apply filters
	if entityType, ok := filter["entity_type"]; ok {
		query = query.Where("entity_type = ?", entityType)
	}
	if entityID, ok := filter["entity_id"]; ok {
		query = query.Where("entity_id = ?", entityID)
	}
	if action, ok := filter["action"]; ok {
		query = query.Where("action = ?", action)
	}
	if userID, ok := filter["user_id"]; ok {
		query = query.Where("user_id = ?", userID)
	}
	if fromDate, ok := filter["from_date"]; ok {
		query = query.Where("created_at >= ?", fromDate)
	}
	if toDate, ok := filter["to_date"]; ok {
		query = query.Where("created_at <= ?", toDate)
	}

	// Count total records
	if err := query.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	// Apply pagination
	offset := (page - 1) * pageSize
	err := query.Order("created_at DESC").Offset(offset).Limit(pageSize).Find(&logs).Error

	return logs, int(total), err
}

func (r *auditLogRepository) GetByEntityID(entityType string, entityID uuid.UUID) ([]domain.AuditLog, error) {
	var logs []domain.AuditLog

	err := r.db.Where("entity_type = ? AND entity_id = ?", entityType, entityID).
		Order("created_at DESC").
		Find(&logs).Error

	return logs, err
}

func (r *auditLogRepository) DeleteOldLogs(beforeDate time.Time) error {
	return r.db.Where("created_at < ?", beforeDate).Delete(&domain.AuditLog{}).Error
}

File: backend/src/repositories/AuditLogRepository.js
const { AuditLog } = require('../models');
const { Op } = require('sequelize');
const logger = require('../utils/logger');

class AuditLogRepository {
  async create(auditData) {
    try {
      const auditLog = await AuditLog.create(auditData);
      return auditLog;
    } catch (error) {
      logger.error('Error creating audit log:', error);
      throw error;
    }
  }

  async list(filter = {}) {
    try {
      const whereClause = this.buildWhereClause(filter);

      const auditLogs = await AuditLog.findAll({
        where: whereClause,
        order: [['created_at', 'DESC']],
      });

      return auditLogs;
    } catch (error) {
      logger.error('Error listing audit logs:', error);
      throw error;
    }
  }

  async listPaginated(filter = {}, page = 1, pageSize = 10) {
    try {
      const offset = (page - 1) * pageSize;
      const whereClause = this.buildWhereClause(filter);

      const { count, rows } = await AuditLog.findAndCountAll({
        where: whereClause,
        order: [['created_at', 'DESC']],
        limit: pageSize,
        offset,
      });

      return {
        data: rows,
        total: count,
        page,
        pageSize,
        totalPages: Math.ceil(count / pageSize),
      };
    } catch (error) {
      logger.error('Error listing audit logs with pagination:', error);
      throw error;
    }
  }

  async getByEntityId(entityType, entityId) {
    try {
      const auditLogs = await AuditLog.findAll({
        where: {
          entity_type: entityType,
          entity_id: entityId,
        },
        order: [['created_at', 'DESC']],
      });

      return auditLogs;
    } catch (error) {
      logger.error('Error getting audit logs by entity ID:', error);
      throw error;
    }
  }

  async deleteOldLogs(beforeDate) {
    try {
      const deletedCount = await AuditLog.destroy({
        where: {
          created_at: {
            [Op.lt]: beforeDate,
          },
        },
      });

      logger.info(`Deleted ${deletedCount} old audit logs before ${beforeDate}`);
      return deletedCount;
    } catch (error) {
      logger.error('Error deleting old audit logs:', error);
      throw error;
    }
  }

  buildWhereClause(filter) {
    const whereClause = {};

    if (filter.entity_type) {
      whereClause.entity_type = filter.entity_type;
    }

    if (filter.entity_id) {
      whereClause.entity_id = filter.entity_id;
    }

    if (filter.action) {
      whereClause.action = filter.action;
    }

    if (filter.user_id) {
      whereClause.user_id = filter.user_id;
    }

    if (filter.from_date && filter.to_date) {
      whereClause.created_at = {
        [Op.between]: [filter.from_date, filter.to_date],
      };
    } else if (filter.from_date) {
      whereClause.created_at = {
        [Op.gte]: filter.from_date,
      };
    } else if (filter.to_date) {
      whereClause.created_at = {
        [Op.lte]: filter.to_date,
      };
    }

    if (filter.search) {
      whereClause[Op.or] = [
        { description: { [Op.iLike]: `%${filter.search}%` } },
        { entity_type: { [Op.iLike]: `%${filter.search}%` } },
        { action: { [Op.iLike]: `%${filter.search}%` } },
      ];
    }

    return whereClause;
  }
}

module.exports = AuditLogRepository;
